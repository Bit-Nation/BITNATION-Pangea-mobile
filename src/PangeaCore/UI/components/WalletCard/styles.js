// @flow

import { StyleSheet } from 'react-native';
import Colors from 'pangea-common-reactnative/styles/colors';

const styles = StyleSheet.create({
  container: {
    marginTop: 5,
    height: 130,
    backgroundColor: Colors.Transparent,
    flex: 1,
    alignItems: 'stretch',
  },
  icon: {
    width: 55,
    marginRight: 5,
    marginTop: 36,
  },
  row: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    borderRadius: 15,
    backgroundColor: Colors.white,
  },
  textColumn: {
    flex: 1,
    marginLeft: 0,
    justifyContent: 'flex-start',
  },
  spacer: {
    flex: 1,
  },
  nameHeading: {
    fontSize: 17,
    color: Colors.BitnationGrayColor,
    fontWeight: 'bold',
    marginBottom: 4,
  },
  nameSubheading: {
    fontSize: 30,
    color: Colors.BitnationHighlightYellowColor,
  },
  button: {
    height: 20,
    borderRadius: 40,
    backgroundColor: Colors.Transparent,
  },
  buttonText: {
    color: Colors.BitnationActionColor,
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginRight: 15,
    marginBottom: 15,
  },
  leftButton: {
    marginRight: 4,
  },
  messageText: {
    color: Colors.BitnationLightColor,
    fontSize: 14,
    marginBottom: 10,
  },
});

export default styles;
