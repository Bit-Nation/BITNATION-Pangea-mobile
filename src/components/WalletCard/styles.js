// @flow

import { StyleSheet } from 'react-native';
import Colors from '../../global/colors';

const styles = StyleSheet.create({

  container: {
    marginTop: 5,
    height: 130,
    backgroundColor: Colors.Transparent,
    flex: 1,
    alignItems: 'stretch',
  },
  icon: {
    width: 55,
    marginRight: 10,
    marginTop: 36,
  },
  row: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    borderRadius: 5,
    backgroundColor: Colors.white,
  },
  textColumn: {
    flex: 1,
    justifyContent: 'flex-start',
  },
  spacer: {
    flex: 1,
  },
  nameHeading: {
    fontSize: 17,
    color: Colors.BitnationGrayColor,
    fontWeight: 'bold',
    marginBottom: 4,
  },
  nameSubheading: {
    fontSize: 30,
    color: Colors.BitnationHighlightYellowColor,
  },
  button: {
    height: 25,
    borderRadius: 14,
    width: 87,
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
  },
  leftButton: {
    marginRight: 8,
  },
  messageText: {
    color: Colors.BitnationLightColor,
    fontSize: 14,
    marginBottom: 10,
  },

});

export default styles;
