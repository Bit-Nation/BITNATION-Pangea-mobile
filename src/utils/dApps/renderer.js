// @flow

import React from 'react';

import Text from '../../components/dApps/Text';
import View from '../../components/dApps/View';
import TextInput from '../../components/dApps/TextInput';
import Button from '../../components/dApps/Button';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => {
  switch (typeName) {
    case 'Text':
      return Text;
    case 'View':
      return View;
    case 'TextInput':
      return TextInput;
    case 'Button':
      return Button;
    default:
      console.warn(`Unknown component type ${typeName}.`);
      return null;
  }
};

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, stateBased: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const stateBasedProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (propName.endsWith('Path') && validProps.stateBased.includes(propName.substring(0, propName.length - 4))) {
      // It's an allowed prop to set or get root component state. That properties are marked by Path suffix.
      stateBasedProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, stateBasedProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = getTypeElementFromText(type);

  if (component === null) {
    return null;
  }

  const { props } = json;

  const { nativeProps = props, stateBasedProps = {}, callbackProps = {} } =
    component.validNativeProps !== undefined
      ? validateProps(json.props, {
        native: component.validNativeProps,
        stateBased: Object.keys(component.stateBasedProps || {}),
        callbacks: component.callbackProps || [],
      }, type)
      : {};

  return React.createElement(
    component,
    { ...nativeProps, ...customPropsProvider(component, { ...nativeProps, ...stateBasedProps, ...callbackProps }), key },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
