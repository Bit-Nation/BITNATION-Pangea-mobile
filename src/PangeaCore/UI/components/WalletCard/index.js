// @flow

import React from 'react';
import { Image, View } from 'react-native';
import { Button, Text } from 'native-base';
import styles from './styles';
import Images from 'pangea-common-reactnative/assets/AssetsImages';
import i18n from 'pangea-common/i18n';

type Props = {
  /**
   * @desc Name for the wallet
   */
  nameHeading: string,
  /**
   * @desc Balance of the wallet
   */
  balance: string,
  /**
   * @desc Image to be displayed
   */
  imagePath: string,
  /**
   * @desc Callback on send money button press
   */
  onSendPress: () => void,
  /**
   * @desc Callback on receive money button press
   */
  onReceivePress: () => void,
  /**
   * @desc Callback on transactions button press
   */
  onTransactionPress: () => void,
};

/**
 * @desc Component for rendering wallet details.
 * @return {React.Component} A component.
 */
const WalletCard = ({
  imagePath,
  nameHeading,
  balance,
  onSendPress,
  onReceivePress,
  onTransactionPress,
}: Props) => (
  <View style={styles.container}>
    <View style={styles.row}>
      <Image style={styles.icon} source={imagePath} resizeMode='contain' />

      <View style={styles.textColumn}>
        <View style={styles.spacer} />
        <Text style={styles.nameHeading}>{nameHeading}</Text>
        <Text style={styles.nameSubheading}>{balance}</Text>
        <View style={styles.spacer} />

        <View style={styles.buttonsContainer}>
          <Button onPress={onSendPress} style={styles.button}>
            <Text style={styles.buttonText}>{i18n.t('common.send')}</Text>
          </Button>
          <Button onPress={onReceivePress} style={styles.button}>
            <Text style={styles.buttonText}>{i18n.t('common.receive')}</Text>
          </Button>
          <Button
            onPress={onTransactionPress}
            style={styles.button}
          >
            <Text style={styles.buttonText}>{i18n.t('common.transactions')}</Text>
          </Button>
        </View>
      </View>
    </View>
  </View>
);

WalletCard.defaultProps = {
  imagePath: Images.ethereumLogo,
  nameHeading: i18n.t('common.ethereum'),
  balance: '0',
  onSendPress: () => null,
  onReceivePress: () => null,
  onTransactionPress: () => null,
};

export default WalletCard;
