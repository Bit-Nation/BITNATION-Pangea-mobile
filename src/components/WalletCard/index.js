// @flow

import React from 'react';
import {
  Text,
  Image,
  View,
} from 'react-native';

import styles from './styles';
import Images from '../../global/AssetsImages';
import Button from '../common/ArrowButton';
import i18n from '../../global/i18n';

type Props = {
  /**
   * @desc Name for the wallet
   */
  nameHeading: string,
  /**
   * @desc Balance of the wallet
   */
  balance: string,
  /**
   * @desc Image to be displayed
   */
  imagePath: string,
  /**
   * @desc Callback on send money button press
   */
  onSendPress: () => void,
  /**
   * @desc Callback on receive money button press
   */
  onReceivePress: () => void,
  /**
   * @desc Callback on transactions button press
   */
  onTransactionPress: () => void,
}

/**
 * @desc Component for rendering wallet details.
 * @return {React.Component} A component.
 */
const WalletCard = ({
  imagePath, nameHeading, balance, onSendPress, onReceivePress, onTransactionPress,
}: Props) => (
  <View style={styles.container}>

    <View style={styles.row}>
      <Image style={styles.icon} source={imagePath} resizeMode='contain' />

      <View style={styles.textColumn}>
        <View style={styles.spacer} />
        <Text style={styles.nameHeading}>{nameHeading}</Text>
        <Text style={styles.nameSubheading}>{balance}</Text>
        <View style={styles.spacer} />

        <View style={styles.buttonsContainer}>
          <Button title={i18n.t('common.send')} onPress={onSendPress} style={[styles.button, styles.leftButton]} />
          <Button title={i18n.t('common.receive')} onPress={onReceivePress} style={styles.button} />
          <Button title={i18n.t('common.transactions')} onPress={onTransactionPress} style={styles.button} />
          <View style={styles.spacer} />
        </View>

        <View style={styles.spacer} />
      </View>
    </View>

  </View>
);

WalletCard.defaultProps = {
  imagePath: Images.ethereumLogo,
  nameHeading: i18n.t('common.ethereum'),
  balance: '0',
  onSendPress: () => null,
  onReceivePress: () => null,
};

export default WalletCard;
