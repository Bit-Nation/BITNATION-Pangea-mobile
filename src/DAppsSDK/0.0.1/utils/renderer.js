// @flow

import React from 'react';
import components from '../components';

export type ComponentsJSON = { type: string, props: Object, children: Array<Object> };

/**
 * @desc Function to get element class or function from text representation, e.g. 'Text' means Text component.
 * @param {string} typeName String representation of the type.
 * @return {*} Class or function of component or null if type is unknown.
 */
export const getTypeElementFromText = (typeName: string) => components[typeName];

/**
 * @desc Validates props against list of valid names and filter only whitelisted ones.
 * Used to prevent DApp developer to pass something bad to components we provide.
 * @param {Object} props Props to validate.
 * @param {string[]} validProps Object containing arrays by keys `native`, `stateBased` and `callbacks` according to corresponding type of props.
 * @param {string} type String representation of component type.
 * @return {Object} Props filtered by groups.
 */
export const validateProps = (
  props: Object,
  validProps: { native: Array<string>, custom: Array<string>, callbacks: Array<string> },
  type: string,
) => {
  const nativeProps = {};
  const customProps = {};
  const callbackProps = {};
  Object.keys(props).forEach((propName) => {
    if (typeof propName !== 'string') {
      console.warn('Prop name is not a string');
      return;
    }

    if (validProps.native.includes(propName)) {
      // It's an allowed native prop.
      nativeProps[propName] = props[propName];
    } else if (validProps.custom.includes(propName)) {
      // It's an allowed custom prop.
      customProps[propName] = props[propName];
    } else if (propName.endsWith('ID') && validProps.callbacks.includes(propName.substring(0, propName.length - 2))) {
      // It's an allowed prop to perform a callback. That properties are marked by ID suffix.
      callbackProps[propName] = props[propName];
    } else {
      console.warn(`Prop ${propName} is not allowed on component ${type}`);
    }
  });

  return { nativeProps, customProps, callbackProps };
};

/**
 * @desc Renders JSON provided by DApp developer into react components.
 * @param {ComponentsJSON} json JSON object that represents components to render.
 * @param {?string} key Key of component (used for updates).
 * @param {Function} customPropsProvider Function that takes component class or function and props and returns new props if it's needed to add ones.
 * @param {*} parent Function or class of component that is parent of currently used or undefined for root.
 * @return {*} Tree of components that is ready to be render.
 */
export const renderJSON = (json: ComponentsJSON, key: ?string, customPropsProvider: (component: any, ownProps: Object) => Object, parent: any) => {
  const { type, children } = json;

  // This is the case when we have string literal inside children.
  // It's only allowed for Text.
  if (type === undefined && typeof (json) === 'string') {
    if (parent === components.Text) {
      return json;
    }

    console.warn(`String literal may only appear inside Text component. Found "${json}" outside.`);
    return null;
  }

  const component = components[type];

  if (component == null) {
    console.warn(`Trying to render unknown component type ${type}.`);
    return null;
  }

  if (component.validNativeProps != null && Array.isArray(component.validNativeProps) === false) {
    console.warn(`validNativeProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.customProps != null && Array.isArray(component.customProps) === false) {
    console.warn(`customProps field of component ${type} should be an array of strings.`);
    return null;
  }
  if (component.callbackProps != null && Array.isArray(component.callbackProps) === false) {
    console.warn(`callbackProps field of component ${type} should be an array of strings.`);
    return null;
  }

  const { props } = json;

  const { nativeProps = props, customProps = {}, callbackProps = {} } =
    validateProps(json.props, {
      native: component.validNativeProps || [],
      custom: component.customProps || [],
      callbacks: component.callbackProps || [],
    }, type);

  return React.createElement(
    component,
    {
      nativeProps,
      ...customProps,
      ...customPropsProvider(component, { ...nativeProps, ...customProps, ...callbackProps }),
      key,
    },
    children
      ? children.map((child, index) => renderJSON(child, `${index}`, customPropsProvider, component))
      : null,
  );
};
