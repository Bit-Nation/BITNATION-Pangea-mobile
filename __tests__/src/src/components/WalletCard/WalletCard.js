import WalletCard from '../../../../../src/components/WalletCard';
import React from 'react';
import renderer from 'react-test-renderer'
import Images from "../../../../../src/global/AssetsImages";

test('WalletCard renders correctly', () => {
  const tree = renderer.create(
    <WalletCard
      imagePath={Images.ethereumLogo}
      nameHeading='Testing Wallet'
      balance='Testing Balance'>
    </WalletCard>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
