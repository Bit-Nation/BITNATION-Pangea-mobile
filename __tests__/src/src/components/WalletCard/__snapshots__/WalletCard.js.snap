// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletCard renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "transparent",
      "flex": 1,
      "height": 130,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-start",
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "marginRight": 10,
          "width": 55,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 17,
            "fontWeight": "bold",
            "marginBottom": 4,
          }
        }
      >
        Testing Wallet
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#72A4DE",
            "fontSize": 14,
          }
        }
      >
        Testing Balance
      </Text>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-start",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 15,
                "height": 30,
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#1C497E",
              },
              Array [
                Object {
                  "backgroundColor": "rgba(114,164,222,0.3)",
                  "borderRadius": 14,
                  "height": 25,
                  "width": 80,
                },
                Object {
                  "marginRight": 8,
                },
              ],
            ]
          }
          title="Send"
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 13,
                "marginRight": 13,
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              buttonTitle={true}
              disabledButtonTitle={false}
              ellipsizeMode="tail"
              style={
                Array [
                  undefined,
                  undefined,
                  Object {
                    "color": "#72A4DE",
                    "fontSize": 14,
                    "letterSpacing": -0.02,
                    "lineHeight": 19,
                    "textAlign": "center",
                  },
                  false,
                  undefined,
                ]
              }
            >
              Send
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "borderRadius": 15,
                "height": 30,
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#1C497E",
              },
              Object {
                "backgroundColor": "rgba(114,164,222,0.3)",
                "borderRadius": 14,
                "height": 25,
                "width": 80,
              },
            ]
          }
          title="Receive"
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "marginLeft": 13,
                "marginRight": 13,
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              buttonTitle={true}
              disabledButtonTitle={false}
              ellipsizeMode="tail"
              style={
                Array [
                  undefined,
                  undefined,
                  Object {
                    "color": "#72A4DE",
                    "fontSize": 14,
                    "letterSpacing": -0.02,
                    "lineHeight": 19,
                    "textAlign": "center",
                  },
                  false,
                  undefined,
                ]
              }
            >
              Receive
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#72A4DE",
            "fontSize": 14,
            "marginBottom": 10,
          }
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
